controller/api/destination/external-workload/endpoints_reconciler.go

```diff
@@ -0,0 +1,656 @@
+		} else {
+			h := epsliceutil.NewPortMapKey(slice.Ports)
+			currentSlicesByPorts[h] = append(currentSlicesByPorts[h], slice)
+		}
+	}
+
+	// Build a list of endpoints we want to create / update. This will be based
+	// off the external workloads we have read.
+	// We use an EndpointSet from the upstream util library. Each Endpoint we
+	// add will be hashed internally.
+	//
+	// We also split each set of endpoints by the ports they target. This
+	// segmentation allows us to have a service select two different workloads
+	// using a named port; consequently there will be 2 endpointslices,
+	// otherwise we'd have a port clash.
+	desiredEndpointsByPort := map[epsliceutil.PortMapKey]epsliceutil.EndpointSet{}
+
+	// A PortMapKey is an upstream type that when created creates a hash of a
+	// port list. We use a map to ensure we don't add ports twice (i.e. we add
+	// them to a set).
+	desiredEndpointPorts := map[epsliceutil.PortMapKey][]discoveryv1.EndpointPort{}
+	for _, extWorkload := range extWorkloads {
+		// We skip workloads with no IPs
+		if len(extWorkload.Spec.WorkloadIPs) == 0 {
+			continue
+		}
+
+		// Find which ports a service selects (or maps to) on an external workload
+		ports, err := r.findEndpointPorts(svc, extWorkload)
+		// Since we require a workload to document its ports, if a service's
+		// targetPort does not match any ports on a workload, we skip it.
+		//
+		// This is different to the upstream where a container port needn't be
+		// documented
+		if err != nil {
+			r.log.Debugf("skipping workload; failed to remap port for service %s/%s and workload %s/%s: %v", svc.Namespace, svc.Name, extWorkload.Namespace, extWorkload.Name, err)
+			continue
+		}
+
+		portHash := epsliceutil.NewPortMapKey(ports)
+		if _, ok := desiredEndpointPorts[portHash]; !ok {
+			desiredEndpointPorts[portHash] = ports
+		}
+
+		ep := externalWorkloadToEndpoint(discoveryv1.AddressTypeIPv4, extWorkload, svc)
+		if _, ok := desiredEndpointsByPort[portHash]; !ok {
+			desiredEndpointsByPort[portHash] = epsliceutil.EndpointSet{}
+		}
+
+		desiredEndpointsByPort[portHash].Insert(&ep)
+	}
+
+	// If there are any slices whose ports no longer match what we want in our
+	// current reconciliation, delete them
+	//
+	// Note: in the upstream they run some more complicated diffing before
+	// applying to ensure creates & deletes turn into updates. We simplify and
+	// instead choose to re-create the slice if ports have changed.
+	for _, currentSlice := range epSlices {
+		portHash := epsliceutil.NewPortMapKey(currentSlice.Ports)
+		if _, ok := desiredEndpointPorts[portHash]; !ok {
+			toDelete = append(toDelete, currentSlice)
+		}
+	}
+
+	totalToCreate := []*discoveryv1.EndpointSlice{}
+	totalToUpdate := []*discoveryv1.EndpointSlice{}
+	totalToDelete := []*discoveryv1.EndpointSlice{}
+
+	// Add slices that we have already marked for deletion due to various
+	// reasons
+	totalToDelete = append(totalToDelete, toDelete...)
```

Comment:
what's the purpose of having a totalToDelete and a toDelete as separate slices? why not just start with a slicesToDelete from the beginning and use it throughout (like in the upstream)?